C/plugins/north/http-north/plugin.cpp

```diff
@@ -0,0 +1,253 @@
+	else if (protocol == string("https"))
+		connector_info->sender = new SimpleHttps(hostAndPort, timeout, timeout);
+	else
+	{
+		Logger::getLogger()->error("Didn't find http/https prefix in URL='%s', cannot proceed", url.c_str());
+		throw new exception();
+	}
+
+	connector_info->path = path;
+	connector_info->proto = protocol;
+	Logger::getLogger()->info("HTTP plugin configured: URL='%s' ", url.c_str());
+
+	return (PLUGIN_HANDLE) (connector_info);
+}
+
+/**
+ * Send Readings data to historian server
+ */
+uint32_t plugin_send(const PLUGIN_HANDLE handle,
+		     const vector<Reading *> readings)
+{
+	CONNECTOR_INFO *connInfo = (CONNECTOR_INFO *) handle;
+	return sendToServer(readings, connInfo);
+}
+
+/**
+ * Shutdown the plugin
+ *
+ * Delete allocated data
+ *
+ * @param handle    The plugin handle
+ */
+void plugin_shutdown(PLUGIN_HANDLE handle)
+{
+	Logger::getLogger()->info("http-north C plugin: %s", __FUNCTION__);
+	CONNECTOR_INFO *connInfo = (CONNECTOR_INFO *) handle;
+	delete connInfo->sender;
+	delete connInfo;
+}
+
+const string& getReadingString(const Reading& reading)
+{
+	string *value = new string;
+	string &m_value = *value;
+
+	// Convert reading data into JSON string
+	m_value.append("{\"timestamp\" : \"" + reading.getAssetDateTime(Reading::FMT_STANDARD) + "Z" + "\"");
+	m_value.append(",\"asset\" : \"" + reading.getAssetName() + "\"");
+	m_value.append(",\"key\" : \"" + reading.getUuid() + "\"");
+	m_value.append(",\"readings\" : {");
+
+	// Get reading data
+	const vector<Datapoint*> data = reading.getReadingData();
+
+	for (vector<Datapoint*>::const_iterator it = data.begin(); it != data.end(); ++it)
+	{
+		m_value.append("\"" + (*it)->getName() + "\": " + (*it)->getData().toString() + ",");
+	}
+
+	// remove last unrequired comma character because of above loop, if one was added
+	auto lastChar = m_value.rbegin();
+	if (*lastChar == ',')
+		m_value.pop_back();
+
+	m_value.append("}}");
+	return m_value;
+}
+
+const vector<pair<string, string>> createMessageHeader()
+{
+	vector<pair<string, string>> res;
+	res.push_back(pair<string, string>("Content-Type", "application/json"));
+	return res; 
+}
+
+uint32_t sendToServer(const vector<Reading *>& readings, CONNECTOR_INFO *connInfo)
+{
+	ostringstream jsonData;
+	jsonData << "[";
+
+	// Fetch Reading* data
+	for (vector<Reading *>::const_iterator elem = readings.begin(); elem != readings.end(); ++elem)
+	{
+		jsonData << getReadingString(**elem) << (elem < (readings.end() -1 ) ? ", " : "");
+	}
+
+	jsonData << "]";
+
+	// Create header for Readings data
+	vector<pair<string, string>> readingData = createMessageHeader();
+
+	// Build a HTTP(S) POST with readingData headers and 'allReadings' JSON payload
+	// Then get HTTPS POST ret code and return 0 to client on error
+	try
+	{
+		int res = connInfo->sender->sendRequest("POST", connInfo->path, readingData, jsonData.str());
+		if (res != 200 && res != 204 && res != 201)
```

Comment:
202 == ACCEPTED, hence should be in too.

Reply:
HTTP status code 202 means "The request has been accepted for processing, but the processing has not been completed. The request might or might not be eventually acted upon, and may be disallowed when processing occurs.". 

So it seems possible that eventually the readings are not ingested at the receiver end. And I think we need to be sure about this here.

Reply:
Its valid code that your request with data has been accepted and server wants to process it later asynchronously. if server is accepting for later execution, we just need to rely on server that it will handle it and ensure data integrity.

>  The request might or might not be eventually acted upon, and may be disallowed when processing occurs.".

It will be bad server, it should have per-processed the data for its type.